Replace Yoast banner with Blaze banner on /stats/ page

DustyReagan opened the pull request:
#### Proposed Changes

* Replace the Yoast SEO banner on /stats with a banner for Blaze.

Before | After
--|--
![stats-banner-before](https://user-images.githubusercontent.com/140841/208198615-4642fe16-316b-4780-a8bb-d945eac333b0.png)  | ![stats-banner-after](https://user-images.githubusercontent.com/140841/208198629-5b4b5207-e86d-4a37-91e8-75a324707906.png)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Load this PR.
* Ensure clicking the banner CTA takes you to `https://wordpress.com/advertising/{blog-domain}`
* Ensure Tracks are being recorded for impressions and clicks
* Make sure banner appears on Simple, Atomic, and Jetpack connected sites
* Ensure the banner matches the Figma. pb5gDS-2fR-p2
* Ensure the secondary Jetpack mobile app banner still functions correctly.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/70835

agrullon95 committed:
Launchpad: Free flow site setup screen (#71254)

* Update free flow site setup step

* Add free flow freeSetup test cases

* Update description placeholder text

jsnmoon committed:
Data Stores: Add handling for undefined site options in getSite (#71283)

WunderBart committed:
E2E: Restore the Page basic flow suite (#70985)

DustyReagan committed:
Add blaze banner to /stats/ page

DustyReagan committed:
Show blaze banner on Jetpack non-Atomic sites

DustyReagan committed:
Update comment

DustyReagan commented on client/my-sites/stats/mobile-promo-cards/index.jsx:
```diff
@@ -2,22 +2,22 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
 	// Mobile apps promos are always shown.
-	const showYoastPromo = ! isOdysseyStats && ! isJetpack;
-	const showBothPromoCards = showYoastPromo;
+	const showBlazePromo = ! isOdysseyStats;
```
Hi @a8ck3n 👋 I noticed you just modified these files. Thought I'd ping you so you know what we're up to. :) Also, I'm honestly not sure what OdyseeyStats is and if we want our Blaze banner showing there or not. Any input here?

a8ck3n replied:
Hey @DustyReagan, thanks for the heads up!

OdysseyStats is the new architecture we are putting in place so we can use the updated Calypso Stats inside wp-admin. The idea being we don't want to code two separate versions of the UI. You can read a bit more about it here: pejTkB-5e

As for these changes, have you looped in @Automattic/wpcom-marketplace so they know about their banner potentially going away? When we (Red Team) touched base with them last, we decided to go with the paged UI so that banner would still be in place.

The changes test well for me on both Atomic & self-hosted Jetpack. One possible change would be to keep the Yoast banner in play but push it to the second page of the DotPager UI. So on a self-hosted site you'd get [Blaze | Jetpack Mobile] while on an Atomic site you'd get [Blaze | Yoast | Jetpack Mobile]. Something to consider at least.

cc: @jeffgolenski @keoshi

cpapazoglou replied:
@DustyReagan can we keep Yoast in the banner as the next item after blaze? 

I can see that Yoast banner has more than 100 clicks daily in tracks dashboard for `calypso_stats_wordpress_seo_premium_banner_click` event.

allilevine replied:
> @DustyReagan can we keep Yoast in the banner as the next item after blaze?
> 
> I can see that Yoast banner has more than 100 clicks daily in tracks dashboard for `calypso_stats_wordpress_seo_premium_banner_click` event.

@cpapazoglou Some context for the Blaze replacement:
- p1668541170792939/1668531558.877439-slack-C029JEQRVRT
- p58i-dut-p2

Leaving Yoast as an item the user can page through to sounds good.

dognose24 committed:
Stats: Polish post detail chart section with period header (#71257)

* Introduce prop showArrows to hide arrows

* Fix icons displaying of StatsSummaryChart

* Polish chart section with the period header

yashwin committed:
Jetpack Pro Dashboard: add toggle to activate monitor on the Jetpack pro dashboard (#71196)

* Create a mutation to update monitor settings

* Create a hook to handle updating monitor settings

* Create a component to toggle activate monitor

* Remove the optimistic update and show notification instead

kangzj committed:
replace odyssey doc link (#71299)

southp committed:
Add: `new` query parameter for the Stepper domain step (#71298)

* Replace the hard-coded `window.location.searchParms` by `useQuery()`

* Allow the Stepper domains step to populate the initial query data from
`new` query parameter.

* For `ref`, use `has()` instead since the actual value doesn't matter

bangank36 committed:
Track if user scrolls to end of Design Picker (#71064)

* Track scroll

* useInView usage

arthur791004 committed:
Design Picker: Fix misalignment by replacing margin-top with row-gap (#71266)

a8ck3n commented on client/blocks/promo-card-block/index.jsx:
```diff
@@ -40,7 +41,7 @@ const PromoCardBlock = ( {
 		<>
 			<QueryJetpackPlugins siteIds={ [ selectedSiteId ] } />
-			{ jetpackNonAtomic || selectedPlugin || isFetching ? (
+			{ ( ! showOnJetpackNonAtomic && jetpackNonAtomic ) || selectedPlugin || isFetching ? (
```
I wonder if it would make more sense to move this business logic outside the component. ie: The caller should determine if the card is relevant and the PromoCardBlock should only be concerned with displaying itself.

In the Stats code, we already do the necessary checks.

paulopmt1 replied:
Sounds good @a8ck3n, [moved its business logic here](https://github.com/Automattic/wp-calypso/pull/71286/commits/e3db9fae1b025865127316bf3da5394325b8a06d)

a8ck3n commented on client/my-sites/stats/mobile-promo-cards/index.jsx:
```diff
@@ -2,22 +2,22 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
 import { translate } from 'i18n-calypso';
 import { useEffect } from 'react';
-import wordpressSeoIllustration from 'calypso/assets/images/illustrations/wordpress-seo-premium.svg';
+import blazeDropDownIllustration from 'calypso/assets/images/illustrations/blaze-drop-down.svg';
```
Just repeating myself but maybe we keep both?

paulopmt1 replied:
Good suggestion. We decided to maintain both banners prioritizing Blaze

kangzj committed:
remove fade in animation for spinner (#71295)

taipeicoder committed:
CSS updates to the theme upsell tooltip (#71264)

msurdi-a8c committed:
Add plans tabs to Pricing Plans block (#71231)

* Add plans tabs to Pricing Plans block

* Add space between left and right sections

* Use separate argument for mandatory class

Co-authored-by: Matías Surdi <matias.emanuel.surdi@automattic.om>

aneeshd16 committed:
BBE Landing: Add localized images for slideshow (#71188)

* Add new images

* Delete old images

* Update images and parameterize animation

* Fixes key prop

wojtekn committed:
Add sites table row component to devdocs (#71195)

* Add SitesTableRow to DevDocs to let us easily test different states

* Use absolute path in example

* Refactor to keep code against DRY rule

* Display all rows in the same table to let us test they together

* Move constants outside the component

* Set displayName for the example component

* Fix SiteThumbnail for devdocs/design as well

ivan-ottinger committed:
Remove Featured image toggle from Writing settings page (#71312)

This commit removes the toggle from the Writing page - as it is already available in the new Reading page.

southp committed:
Do not remove the user step if it's the only one in the flow. (#71302)

ivan-ottinger committed:
Remove `settings/featured-image-template-toggle` feature flag (#71314)

The commit removes the flag that is no longer necessary.

karenroldan committed:
E2E: Fix flaky test - plugin - install/remove (#71082)

* E2E: Fix plugin install/remove

* E2E: Update encrypted secret file for user to newer schema

* E2E: Rename method and update comment

* E2E: Define active property in plugin params

* E2E: Updaate message

* E2E: Update data helper to get site URL on new schema

sirbrillig committed:
format-currency: Fix IDR locale formatting (#71340)

* Add formatCurrency test for IDR

* Fix precision on IDR currency in format-currency

jjolmo committed:
Handle redirecting on users without sites (#71338)

Co-authored-by: Javier Olmo <javier.olmo@automattic.com>

danielbachhuber committed:
Use a 'Pending' icon for SSH keys because they're optional (#71321)

nateweller committed:
Add core version threat type (#66245)

valterlorran committed:
Making some props more clear (#71317)

paulopmt1 committed:
Stats: Some cleanup of PromoCards wrapper component (#71259)

* Renames the MobilePromoCardWrapper component to PromoCards
* Renames the containing directory to match component name
* Cleans up the display logic and HTML output a bit

paulopmt1 committed:
Merge branch 'trunk' into add/stats-blaze-banner

paulopmt1 commented:
Fixed merge conflicts, and working on [removing business logic from it](https://github.com/Automattic/wp-calypso/pull/71286#discussion_r1051865470).

paulopmt1 committed:
Removed business logic from the component

paulopmt1 commented on client/blocks/promo-card-block/index.jsx:
```diff
@@ -40,7 +35,7 @@ const PromoCardBlock = ( {
 	return (
 		<>
 			<QueryJetpackPlugins siteIds={ [ selectedSiteId ] } />
-			{ jetpackNonAtomic || selectedPlugin || isFetching ? (
```
We removed the business logic from the component as /status page is already doing it. [The only other place](https://github.com/Automattic/wp-calypso/blob/add/stats-blaze-banner/client/my-sites/marketing/traffic/index.js#L53) which is reusing this component[ will have Blaze banner](https://github.com/Automattic/wp-calypso/pull/71276) on it, so this logic can be safely removed from here.

paulopmt1 commented on client/my-sites/stats/promo-cards/index.jsx:
```diff
@@ -2,49 +2,70 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
-	const showYoastPromo = ! isOdysseyStats && ! isJetpack;
-	// Handle initial view event if not using the DotPager UI.
-	// We don't worry about the Yoast card as it sends on mount.
-	useEffect( () => {
```
Replaced this logic [with a simpler one](https://github.com/Automattic/wp-calypso/pull/71286/files#diff-d0adee9cff76fc082d6cdc6ebfa2dde684cfa28d52007d8c69599eb26b943f87R47).

paulopmt1 commented on client/my-sites/stats/promo-cards/index.jsx:
```diff
@@ -2,49 +2,70 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
+					{ showYoastPromo && (
 						<PromoCardBlock
 							productSlug="wordpress-seo-premium"
-							impressionEvent={ EVENT_YOAST_PROMO_VIEW }
```
We already register it on `pagerDidSelectPage`

paulopmt1 reviewed (approved):
It is working great @DustyReagan! I did some adjustments as we talked.

I can confirm on a self-hosted site we get [Blaze | Jetpack Mobile] while on an Atomic|Simple sites we get [Blaze | Yoast | Jetpack Mobile]

a8ck3n commented on client/my-sites/stats/promo-cards/index.jsx:
```diff
@@ -2,49 +2,70 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
 import './style.scss';
 
 const EVENT_YOAST_PROMO_VIEW = 'calypso_stats_wordpress_seo_premium_banner_view';
+const EVENT_BLAZE_PROMO_VIEW = 'calypso_stats_blaze_banner_view';
```
I'm not sure if you wish to promote Blaze on other Stats pages but if you do, you might want the identifier to account for that. Depending on how the designs come out, we might also have promo cards on the Insights, Ads, or Store pages. See the note inside `promoCardDidReceiveClick` where it recommends a format of `calypso_stats_PAGE_mobile_cta_EVENT`.

paulopmt1 replied:
That sounds good! Changed. Combined [with the other page event](https://github.com/Automattic/wp-calypso/pull/71276/files#diff-368bfab8cff4892b6168822c3412ce9199f2a159befb6032b25fda7b0bb0887fR55), it should work pretty well!

a8ck3n commented on client/my-sites/stats/promo-cards/index.jsx:
```diff
@@ -2,49 +2,70 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
 	const pagerDidSelectPage = ( index ) => {
-		const evenLookup = [ EVENT_YOAST_PROMO_VIEW, EVENT_MOBILE_PROMO_VIEW ];
-		const eventName = evenLookup[ index ];
+		const eventLookup = [ EVENT_BLAZE_PROMO_VIEW, EVENT_YOAST_PROMO_VIEW, EVENT_MOBILE_PROMO_VIEW ];
```
This logic no longer works now that the DotPager is variable. For example, if you visit a self-hosted site and view the Jetpack Mobile card, it will send the Yoast event. This array needs to be updated based on the same state used for the DotPager pages.

paulopmt1 replied:
Good catch!

a8ck3n commented on client/my-sites/stats/promo-cards/index.jsx:
```diff
@@ -2,49 +2,70 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
-	// Render one or both promo cards.
+
+	// Trigger initial view event as DotPager don't call it.
+	useEffect( () => pagerDidSelectPage( 0 ) );
```
This assumes the DotPager resets to page zero when changing sites which isn't the case.

1. Select a site and view the 2nd page (Yoast).
2. Changing sites will trigger a new Blaze view event despite the card still showing Yoast.

The old code handled this by not sending new impression events when the site changed which seems reasonable. This new behaviour will send the wrong impression events though.

It's a limitation of the DotPager that it doesn't allow us to reset the page so the correct fix might be to track the last selected page as local state and use that, in combo with the `slug`, as dependancies for firing this.

paulopmt1 replied:
Thanks for all the tests and this fix suggestion @a8ck3n!

I tried to fix storing the last selected page as local state combining with slug change but didn't find a good fix in my timebox.

It seems the best way to track this would be to fix DotPager's limitation by resetting the view, so I brought back the old logic of not sending new impressions events when the site changed, as we already sent them on the component mount.

a8ck3n reviewed (changes requested):
I'm seeing a few cases where the analytics are no longer correct.

The case of firing the Yoast event when it should be a Jetpack Mobile event should be addressed before merging. (see second comment below)

I'm less sure about the impression events when changing sites (3rd comment). My thinking is we could skip sending these (are they really new impressions?) but if we wanted to record them, that probably means tracking the selected page per my comment below.

a8ck3n commented:
> Fixed merge conflicts, and working on [removing business logic from it](https://github.com/Automattic/wp-calypso/pull/71286#discussion_r1051865470).

You should be able to avoid the polluted commit history (and excess GH pings) by using `--force-with-lease` per this comment: p7H4VZ-3Oh#comment-10717

paulopmt1 committed:
Merge branch 'trunk' into add/stats-blaze-banner

paulopmt1 committed:
Events rearranged to match elements in DotPager

paulopmt1 committed:
Improved Blaze event name

paulopmt1 committed:
Improved Blaze click event name

lupus2k reviewed (approved):
I tested en Atomic / Simple / Jetpack sites, and all banners work perfectly. 
Good job 🎉

Commit:
Fix logic for firing view events

jsnmoon reviewed (changes requested):
Blaze promo view and Yoast promo view events currently fire indiscriminately (and incorrectly) upon mounting. This is caused by `PromoCardBlock`'s `impressionEvent`, which fires upon mounting the component whether or not it's made visible through the DotPager.

<img width="1384" alt="image" src="https://user-images.githubusercontent.com/4044428/209221022-7898a98c-9496-4bac-b09b-41c579d8adeb.png">

Note the timestamps between the page view event and the two banner events. This log output is created by the `debug` package and can be enabled by setting `localStorage.debug = "calypso:analytics*"` in your browser console.

This PR has sat in code review long enough, so I went ahead and fixed this issue (as well as the selected site switcher issue) in [a separate branch here](https://github.com/Automattic/wp-calypso/compare/add/stats-blaze-banner...add/stats-blaze-banner-with-tracks-fix). Feel free to cherry-pick or riff on my changes for this PR.

paulopmt1 committed:
Fixed out of bounds index when switching sites

paulopmt1 commented:
> Blaze promo view and Yoast promo view events currently fire indiscriminately (and incorrectly) upon mounting. This is caused by `PromoCardBlock`'s `impressionEvent`, which fires upon mounting the component whether or not it's made visible through the DotPager.
> 
> <img alt="image" width="1384" src="https://user-images.githubusercontent.com/4044428/209221022-7898a98c-9496-4bac-b09b-41c579d8adeb.png">
> 
> Note the timestamps between the page view event and the two banner events. This log output is created by the `debug` package and can be enabled by setting `localStorage.debug = "calypso:analytics*"` in your browser console.
> 
> This PR has sat in code review long enough, so I went ahead and fixed this issue (as well as the selected site switcher issue) in [a separate branch here](https://github.com/Automattic/wp-calypso/compare/add/stats-blaze-banner...add/stats-blaze-banner-with-tracks-fix). Feel free to cherry-pick or riff on my changes for this PR.

Thanks for your help @jsnmoon! It's working correctly!
I added extra checking for switching from an Atomic site (3 banners) to a jetpack connected (2 banners) and not having issues with index boundaries.

I feel the next step to avoid all these codes in future uses of `DotPager` is to fix its behavior to trigger the view event properly by having a `onPageView` event. How do you see it?

paulopmt1 commented:
I am merging this as the code is stable and meets the goal. Any enhancement can be done on follow-up tasks.
Thanks for your help!

jsnmoon commented:
> I am merging this as the code is stable and meets the goal. Any enhancement can be done on follow-up tasks.
Thanks for your help!

Nice! Happy to help 🙂 

> I feel the next step to avoid all these codes in future uses of DotPager is to fix its behavior to trigger the view event properly by having a onPageView event. How do you see it?

Dynamically mounting visible components sounds like a great enhancement to DotPager. We'll want to take extra care in preserving the paging animation (contents sliding in & out), which currently assumes all of DotPager's children to be mounted.
